sequencer.c

```diff
@@ -3848,6 +3859,26 @@ static int pick_commits(struct repository *r,
 				setenv("GIT_REFLOG_ACTION", reflog_message(opts,
 					command_to_string(item->command), NULL),
 					1);
+			if (peek_command(todo_list, 1) == TODO_PICK &&
+				is_fixup(peek_command(todo_list, 2))) {
+				char *p;
+
+				if (get_oid("HEAD", &head))
+					return error(_("cannot read HEAD"));
+
+				p = oid_to_hex(&head);
+				write_message(p, strlen(p), rebase_path_bad_fixup(), 1);
```

Comment:
Good idea, but I would actually write this only in case of a failed `TODO_PICK`, `TODO_EDIT` or `TODO_REWORD`, i.e. if `is_pick_or_similar()` returns `1` (but make sure that we don't do it if `is_fixup()` returns `1`).

Further, I think that `peek_command(todo_list, 1)` is looking too far in the future. We are about to call `do_pick_commit()` with `item->command`, so that's the command to look at. Besides, at this point `peek_command(todo_list, 0)` already looks at the next command.

So I think that the logic should go more like this:

```c
struct strbuf buf = STRBUF_INIT;
struct object_id head, previous_head;

[...]
res = 0;
if (is_fixup(item->command)) {
    if (read_oneliner(&buf, rebase_path_avoid_bad_fixup(), 1) > 0) {
        if (get_oid(buf.buf, &previous_head)) {
            error(_("invalid file contents '%s' of '%s'"),
                  buf.buf, rebase_path_avoid_bad_fixup());
            strbuf_release(&buf);
        }
        strbuf_release(&buf);
        if (get_oid("HEAD", &head))
            return error(_("could not read HEAD"));
        if (oideq(&head, &previous_head))
            res = error(_("commit to fix up was skipped")); /* reschedule */
    }
} else if (is_pick_or_similar(item->command) &&
            is_fixup(peek_command(todo_list, 0)) {
    char hex[GIT_MAX_HEXSZ + 1];

    if (get_oid("HEAD", &head))
        return error(_("cannot read HEAD"));
    oid_to_hex_r(hex, &head);
    write_message(hex, the_hash_algo->hexsz, rebase_path_avoid_bad_fixup(), 0);
}

if (!res)
    res = do_pick_commit(r, item->command, item->commit,
					     opts, is_final_fixup(todo_list),
					     &check_todo);
```

Reply:
I still have 2 questions right now,

- Is there any particular advantage in writing to file, since we are only writing a single value ?
- What does oid_to_hex_r do differently ?

Reply:
`struct strbuf buf = STRBUF_INIT;`
 And how about reducing this variable's scope even further ? Into


> ```c
> if (is_fixup(item->command)) {
>     if (read_oneliner(&buf, rebase_path_avoid_bad_fixup(), 1) > 0) {
>         if (get_oid(buf.buf, &previous_head)) {
>             error(_("invalid file contents '%s' of '%s'"),
>                   buf.buf, rebase_path_avoid_bad_fixup());
>             strbuf_release(&buf);
>         }
>         strbuf_release(&buf);
>         if (get_oid("HEAD", &head))
>             return error(_("could not read HEAD"));
>         if (oideq(&head, &previous_head))
>             res = error(_("commit to fix up was skipped")); /* reschedule */
>     }
> }
> ```

Reply:
The value is short all right, but remember: the most likely scenario is that the `pick` *failed* due to merge conflicts, the user `--skip`s it, and *then* the `fixup` has no correct commit to target.

In other words: `git` stopped running between writing the value and reading it. Those are two separate processes, therefore the value *must* be persisted to disk.

About reducing the scope even further: sure, as long indentation issues don't make the code unreadable?